### ISSUE TITLE
PM981 hangs on install

### ISSUE THREAD

**Issue by tane123** [acidanthera/bugtracker#841]
Hi there, I am currently trying to get A pm981 ssd to install catalina 10.15.4. Currently I have nvmefix 1.0.2 and lilu 1.4.3 installed with the patch listed in another issue comment #780 but when to install it reboots once. Starts installing but comes up with Mac OS catalina cannot be installed safely. I can provide anything needed. I tried installing the debugging version of Open core but I cant seen to get into OC with the latest version. I tried to convert my config.plist to the latest version  but OpenCore itself wouldn't boot.

**Comment by 07151129** [comment]
You need to get OpenCore working properly first.

> Starts installing but comes up with Mac OS catalina cannot be installed safely.

What does this mean?

**Comment by sinag** [comment]
I did try with latest version of nvmefix and lilu etc. 
My laptop is dell xps 9380 with pm981 512gb.
I could install 10.15.4 to a secondary usb hard disk.
Then I tried to install on pm981. Started clean with a single macos partition. Created a macos journalled single partition. Installer failed at 2minutes remaining and reboots when trying to convert hfs+ to afps.
Can also reproduce without installing. Create a hfs partition and try to convert to afps from diskutil, or create apfs at the beginning, crash.

**Comment by tane123** [comment]
Sorry for not being clear. I managed to get other debug version of opencore working as well as a working log.  The installer will halt at 13 minutes remain, attached is my log file, full efi folder and a folder of the point at which it stops installling. I also can not use Diskutil to erase my volumes, it crashes at 38% whilst creating the partition map. Anything you want me to provide I can. 
[NVMe debug.zip](https://github.com/acidanthera/bugtracker/files/4467736/NVMe.debug.zip)

**Comment by sinag** [comment]
I can confirm similar results as @tane123 
I can create a hfs+ partition using diskutil, converting to apfs fails
On a usb disk, I could create apfs partition without a problem

**Comment by 07151129** [comment]
@tane123 So does it halt or does it reboot?

The information you provided is not helpful. Can you send kernel log instead?

**Comment by tane123** [comment]
It just halts. Not reboots no visible kernel panics or anything of that sort. It starts to install and halts on 13 minutes remaining.  I'm not quite sure what you mean by kernel log or how to get it/
enable it? In the zip folder is a Opencore log if that is what you mean? 
![93203819_300799284238513_7869355490524790784_n](https://user-images.githubusercontent.com/19453292/79180865-1fe02e80-7e5f-11ea-8473-bd9c9abce8b4.jpg)
[opencore-2020-04-12-233348.txt](https://github.com/acidanthera/bugtracker/files/4472850/opencore-2020-04-12-233348.txt)

**Comment by tane123** [comment]
Its halts. There is no reboots or kernel panics of any sort. Im not sure how to access the kernel log? Its is stored in the nvram or something? The only log I have is the opencore log file. 
[opencore-2020-04-14-040232.txt](https://github.com/acidanthera/bugtracker/files/4473049/opencore-2020-04-14-040232.txt)

**Comment by sinag** [comment]
@tane123 do you have a repo where you track your opencore or clover efi? What is your hardware?

**Comment by 07151129** [comment]
> I can create a hfs+ partition using diskutil, converting to apfs fails

Fails with what error?

**Comment by armenio** [comment]
@07151129 i'm able to provide any necessary information.

NVMe PM981 is so problematic that I had to install the OS on a USB stick on a second machine as it gives KP timeout randomly even during an installation that does not involve it.

when I boot from the pendrive (external), I need to quickly eject the PM981, this avoids KP and I am able to work normally for hours using the pendrive boot.

I am not familiar with providing the necessary data to solve this problem, but I have technical skills as a programmer and if you give me the instructions I will do my best to contribute.

your work and dedication is fantastic and I really want to contribute as much as possible to make that happen.

aside, is formatting the disk in block size 4k with linux nevme-cli an option for it to work?

**Comment by tane123** [comment]
Okay so I managed to get the kernel panic that was causing the problem. Attached is that error also a few other logs. I tried finding a kernel log in a few system locations but only found the system and kernel panic.  Attached is both but i doubt the system log will be helpful. Also i agree with @armenio, you dedication to this project is amazing. Thanks for you support so far. 

[Debug.zip](https://github.com/acidanthera/bugtracker/files/4478185/Debug.zip)

**Comment by sinag** [comment]
> > I can create a hfs+ partition using diskutil, converting to apfs fails
> 
> Fails with what error?

You are more than right to ask for usefull feedback, and I agree with @tane123 and @armenio that I appreciate your work on this.

I am not experienced with kernel development so I need to learn how to get usefull logs first. As I dont have a spare computer, each time I try installing macos, I completely kill my Windows partition thinking it could be the cause of a problem which means killing the windows os where I mainly do my daily work, so after that I reinstall Windows and everything which is hard work. So next time I try this, I will try to get and capture more usefull data.

Any link, help, guide on getting required logs will be highly appreciated btw.
Thx for your work.

**Comment by vit9696** [comment]
I guess I will close this since we will unlikely investigate PM981 issues we cannot reproduce on our hardware. Sorry for the false hopes.

**Comment by sinag** [comment]
I have managed to get some logs finally if it makes sense.
[system.log](https://github.com/acidanthera/bugtracker/files/4491449/system.log)
Sometimes it hangs on installer ui when creating the partition, sometimes during installation to newly created partition.

**Comment by 07151129** [comment]
Sadly, this still does not help: the log contains no NVMe-related errors.

Can you try test creating APFS or HFS+ partitions by hand (using `newfs_apfs`, `newfs_hfs`)?

**Comment by sinag** [comment]
Sure I can do that.
But, when I do disk related things, creating apfs partition at diskutil or writing to this partition 
 during installation, system freezes on user interface even if I have keep symbols and dont reboot on panic at clover. 
At that point how can I collect log files? Is there a feature on clover to automatically create these files? The log that I sent was from "/var/log/system.log" which I have copied to efi partition by hand when I was able to use the system with terminal.

**Comment by tane123** [comment]
Alright so i just had a go with newfs_apfs but no cigar. I did manage to pull my boot logs surrounding nvme. Not sure if anything sticks, I am currently using log stream to follow the kernel to see at which code it crashes at. Ill get back to you once it has crashed. Attached is my kernel log for nvme debug will post with live kernel log soon.
[Kernel debug nvme.txt](https://github.com/acidanthera/bugtracker/files/4493246/Kernel.debug.nvme.txt)

**Comment by tane123** [comment]
Here are some more log files of which does have nvme entries thank god!
[Kernel log before crash.txt](https://github.com/acidanthera/bugtracker/files/4493305/Kernel.log.before.crash.txt)
[full kernel log.txt](https://github.com/acidanthera/bugtracker/files/4493307/full.kernel.log.txt)

**Comment by 07151129** [comment]
@tane123 can you check which symbols it fails to solve using DEBUG build of Lilu?

**Comment by tane123** [comment]
Attached is a full lilu debug and nvme debug with the lilu 
[nvme debug.txt](https://github.com/acidanthera/bugtracker/files/4498488/nvme.debug.txt)
[lilu debug.txt](https://github.com/acidanthera/bugtracker/files/4498489/lilu.debug.txt)

**Comment by tane123** [comment]
From what I can see in the logs i think its related to getOSData vendor-id was not found? Im not sure thats just a guess

**Comment by 07151129** [comment]
@tane123 how did you install NVMeFix? Are you sure you have only one instance of IONVMeFamily (in /System/Library/Extensions, and not elsewhere)? What is `md5` of IONVMeFamily binary (in IONVMeFamily.kext/Contents/MacOS/IONVMeFamily)?

**Comment by tane123** [comment]
Im injecting it using the bootloader. I checked for double instances and nothing. Only the one instance. The MD5 hash is b485301e1d806239358020970e571130

**Comment by tane123** [comment]
Any luck with the NVMEfix with extra debugging? Do you think installing it into S/L/E would help at all?

**Comment by 07151129** [comment]
> Any luck with the NVMEfix with extra debugging? Do you think installing it into S/L/E would help at all?

https://github.com/acidanthera/NVMeFix/commit/9d72ff387bb194f9ea00f344382abe13bf5d1d44 adds extra logging for DEBUG builds.



> The MD5 hash is b485301e1d806239358020970e571130

Which OS build is that? I checked  10.15.4 (19E287), and it has e494d04297ed7744df8e572e26ccd198

**Comment by tane123** [comment]
Attached is the debug logs with extra logging. It was 10.15.0 but I updated to 10.15.4 now the md5 matches. e494d04297ed7744df8e572e26ccd198. Still seems the same issue? 
[Debug.zip](https://github.com/acidanthera/bugtracker/files/4534404/Debug.zip)

**Comment by 07151129** [comment]
Something (bootloader patch?) is interfering, because NVMeFix is looking for the symbol correctly.

**Comment by tane123** [comment]
Alright so I managed to install and boot Mac Os 10.15.4 but after a few reboots it starts to kernel panics and it randomly boots on ocasions.  Attached is the kernel panic and logs on successful boot . It seems to be related to when the kernel tries and opens the corefile? I tried disabling core dumps but no luck. Is there anyway to pull a kernel log from a non-booting systems by using a removable boot disk? Also when you guys were testing did you use any particular boot flags or device properties? The only reason i managed to get mac os installed was because I hadnt disabled apst. I honestly have no idea what is interfering with it being able to solve the symbols.  
![20200429_133355](https://user-images.githubusercontent.com/19453292/80600005-e5da7380-8a7f-11ea-9899-767a5c528c56.jpg)
![20200429_134158](https://user-images.githubusercontent.com/19453292/80600008-e7a43700-8a7f-11ea-828a-4c13a217b1fd.jpg)
[debug.zip](https://github.com/acidanthera/bugtracker/files/4552200/debug.zip)
